### DISCUSSION TITLE
关于前台脚本交互

### DISCUSSION THREAD

**Discussion by jc3213** [jc3213/download_with_aria2#51] [Ideas]
关于这个部分，一切以[Wiki:Message](https://github.com/jc3213/download_with_aria2/wiki/Message)为准

**Comment by jc3213** [comment]
@ivysrono @wordpure @daiaji

**Reply by daiaji** [reply]
aria2c改了这玩意？

**Reply by jc3213** [reply]
这个只单纯的加入了之前讨论提到的交互内容，ctrl+alt 快速下载你只需要把之前shortcut.js的内容复制进来，把**type**改成`download`就行了。

我只是觉得自己实现的方案很蠢，因此需要FDM或者IDM的浏览器扩展源代码参考。如果可以的话麻烦提供一下吧OTL。

**Reply by daiaji** [reply]
[IDM_CRX.zip](https://github.com/jc3213/download_with_aria2/files/10154755/IDM_CRX.zip)
三个CRX,默认IDM的行为是按住DELETE键就不捕获下载请求。
设置界面在IDM里，不知道是在系统的全局热键里实现还是在扩展里实现。

**Reply by jc3213** [reply]
我昨天搜索了下官方的扩展，但是全是简化代码，我直接麻了。不过整体研究过了还算能看懂，我就看到默认抓取，而且是识别`event.target.localName`。问题是类似`<a href="aaa.com"><i class="icon">some.jpg</i></a>`这种直接用click事件监听并不会得到`a`这个结果。因为是简化代码我没法子看他怎么做的相关的work around，头皮发麻。我自己的方案是额外检查一次`event.target.parentNode`。目前试用还算成功？

对了，可以的话帮忙修复一下linux的build脚本吧，我不太懂sh……

**Reply by daiaji** [reply]
啊这，还真是所有扩展都做了热键绑定，挺肝的。
shell那边我之后再看看。
我得把我的主目录备份了。
都被数据丢失搞怕了。

**Reply by jc3213** [reply]
额，你的储存介质不会是鱼竿，海康之类的吧。OTL

**Reply by daiaji** [reply]
事及其高贵的三爽gen4企业级旗舰pm1733，我安全擦除时忘记确定接在主板上的是哪个SSD了，然后就没了。

**Comment by jc3213** [comment]
同时更新了下[讨论区的脚本](https://github.com/jc3213/download_with_aria2/discussions/37)

**Comment by ivysrono** [comment]
不用IDM，也很久没用FDM了。
下载B站目前看到最方便的脚本：https://greasyfork.org/zh-CN/scripts/413228

**Reply by jc3213** [reply]
没卵用，有下载限制，搞得花里胡哨本质功能跟我那个差不多。而且首选flv（单文件），我的界面还更简洁，还能选编码格式，最重要的是这个依然不能破解会员下载。我还不如直接用youtube-dl了。

下载av1编码，视频+音频容量很小。而且这么简单的功能至于做出200多K的超级脚本吗……而且我看反馈里面也在反应解析flv地址失败了。

**Reply by ivysrono** [reply]
这个脚本最舒服的是批量下载播放列表。

**Reply by jc3213** [reply]
我基本没有批量下载需求……我现在最想知道的是其他工具怎么实现的按住XX键点链接直接下载的……

**Reply by ivysrono** [reply]
偶尔单个下载用啥工具都行，我常用 https://github.com/iawia002/lux/releases
批量下载就很麻烦了。

**Reply by jc3213** [reply]
我的大多数工具都是以实现我自己的需求为前提制作的……没需求就不管了，233。现在主要是这个接口，通过window.postMessage实现跟扩展的交互，不知道你那边测试正常与否。反正讨论组里的脚本，以及我自己写的哔哩哔哩视频下载脚本倒是没问题

**Comment by jc3213** [comment]
支持了批量下载，比如 `{url: [array], options: object}`
```javascript
postMessage({ aria2c: "Download With Aria2", type: "download", message: { url: ['https://github.com/jc3213/download_with_aria2/archive/refs/tags/4.1.0.1612.zip', 'https://github.com/jc3213/download_with_aria2/archive/refs/tags/4.0.1.1504.zip'], options: {dir: 'R:\\'} } })
```

目前，还没法针对`[ {url, options}, {url, options} ]` 进行下载。也就意味着每个文件如果要独立下载目录，下载选项，文件名的话将无法实现。

**Reply by jc3213** [reply]
我现在在想怎么传递`JSON`类别。额外增加个变量还是啥。反正很难顶。

**Comment by jc3213** [comment]
改了下逻辑，现在支持了json，但是所有接收信息的操作必定弹出下载框确认。
例子
```javascript
postMessage({ aria2c: "Download With Aria2", type: "download", message: { json: [ {url: 'https://github.com/jc3213/download_with_aria2/archive/refs/tags/4.1.0.1612.zip', options: {out: '4.1.0.1612.zip'}} , {url: 'https://github.com/jc3213/download_with_aria2/archive/refs/tags/4.0.1.1504.zip', options: {out: '4.0.1.1504.zip'}} ] } })
```
![1](https://user-images.githubusercontent.com/8744936/206067753-0dfdb673-dd6f-4761-9209-822fea66c92e.png)
![2](https://user-images.githubusercontent.com/8744936/206067884-c13b1528-cd90-4501-9017-05401271bdb5.png)

**Comment by jc3213** [comment]
更新了wiki相关内容 https://github.com/jc3213/download_with_aria2/wiki/Message
